File: backend/src/main/java/com/mobileparts/service/CartService.java
package com.mobileparts.service;

import com.mobileparts.entity.Component;
import com.mobileparts.repository.ComponentRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import java.io.Serializable;
import java.math.BigDecimal;
import java.util.*;
import java.util.concurrent.TimeUnit;

/**
 * Redis-based cart service
 * Stores cart items in Redis for fast access and automatic expiration
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class CartService {

    private final RedisTemplate<String, Object> redisTemplate;
    private final ComponentRepository componentRepository;

    private static final String CART_KEY_PREFIX = "cart:";
    private static final long CART_EXPIRATION_HOURS = 24; // Cart expires after 24 hours

    /**
     * Cart item DTO stored in Redis
     */
    public static class CartItemDTO implements Serializable {
        private static final long serialVersionUID = 1L;
        
        private Long id;  // This is the component ID (for backwards compatibility with frontend)
        private Long componentId;
        private String componentName;
        private BigDecimal price;
        private String imageUrl;
        private Integer quantity;

        public CartItemDTO() {}

        public CartItemDTO(Long componentId, String componentName, BigDecimal price, String imageUrl, Integer quantity) {
            this.id = componentId;
            this.componentId = componentId;
            this.componentName = componentName;
            this.price = price;
            this.imageUrl = imageUrl;
            this.quantity = quantity;
        }

        // Getters and setters
        public Long getId() { return id; }
        public void setId(Long id) { this.id = id; }
        
        public Long getComponentId() { return componentId; }
        public void setComponentId(Long componentId) { this.componentId = componentId; }
        
        public String getComponentName() { return componentName; }
        public void setComponentName(String componentName) { this.componentName = componentName; }
        
        public BigDecimal getPrice() { return price; }
        public void setPrice(BigDecimal price) { this.price = price; }
        
        public String getImageUrl() { return imageUrl; }
        public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }
        
        public Integer getQuantity() { return quantity; }
        public void setQuantity(Integer quantity) { this.quantity = quantity; }
    }

    /**
     * Get cart key for user
     */
    private String getCartKey(Long userId) {
        return CART_KEY_PREFIX + userId;
    }

    /**
     * Add item to cart
     */
    public CartItemDTO addToCart(Long userId, Long componentId, Integer quantity) {
        Component component = componentRepository.findById(componentId)
            .orElseThrow(() -> new RuntimeException("Component not found"));

        String cartKey = getCartKey(userId);
        String itemKey = componentId.toString();

        // Check if item already exists in cart
        CartItemDTO existingItem = (CartItemDTO) redisTemplate.opsForHash().get(cartKey, itemKey);

        CartItemDTO cartItem;
        if (existingItem != null) {
            // Update quantity
            existingItem.setQuantity(existingItem.getQuantity() + quantity);
            cartItem = existingItem;
        } else {
            // Create new cart item
            cartItem = new CartItemDTO(
                component.getId(),
                component.getName(),
                component.getPrice(),
                component.getImageUrl(),
                quantity
            );
        }

        // Save to Redis
        redisTemplate.opsForHash().put(cartKey, itemKey, cartItem);
        
        // Set expiration
        redisTemplate.expire(cartKey, CART_EXPIRATION_HOURS, TimeUnit.HOURS);

        log.info("Added {} x {} to cart for user {}", quantity, component.getName(), userId);
        return cartItem;
    }

    /**
     * Get all cart items for user
     */
    public List<CartItemDTO> getCartItems(Long userId) {
        String cartKey = getCartKey(userId);
        Map<Object, Object> cartMap = redisTemplate.opsForHash().entries(cartKey);
        
        List<CartItemDTO> items = new ArrayList<>();
        for (Object value : cartMap.values()) {
            items.add((CartItemDTO) value);
        }
        
        return items;
    }

    /**
     * Update cart item quantity
     */
    public CartItemDTO updateCartItemQuantity(Long userId, Long componentId, Integer quantity) {
        String cartKey = getCartKey(userId);
        String itemKey = componentId.toString();

        CartItemDTO cartItem = (CartItemDTO) redisTemplate.opsForHash().get(cartKey, itemKey);
        
        if (cartItem == null) {
            throw new RuntimeException("Cart item not found");
        }

        if (quantity <= 0) {
            // Remove item if quantity is 0 or negative
            redisTemplate.opsForHash().delete(cartKey, itemKey);
            return null;
        }

        cartItem.setQuantity(quantity);
        redisTemplate.opsForHash().put(cartKey, itemKey, cartItem);
        
        // Refresh expiration
        redisTemplate.expire(cartKey, CART_EXPIRATION_HOURS, TimeUnit.HOURS);

        return cartItem;
    }

    /**
     * Remove item from cart
     */
    public void removeFromCart(Long userId, Long componentId) {
        String cartKey = getCartKey(userId);
        String itemKey = componentId.toString();
        redisTemplate.opsForHash().delete(cartKey, itemKey);
        
        log.info("Removed component {} from cart for user {}", componentId, userId);
    }

    /**
     * Clear entire cart
     */
    public void clearCart(Long userId) {
        String cartKey = getCartKey(userId);
        redisTemplate.delete(cartKey);
        
        log.info("Cleared cart for user {}", userId);
    }

    /**
     * Get cart item count
     */
    public Integer getCartItemCount(Long userId) {
        String cartKey = getCartKey(userId);
        Long size = redisTemplate.opsForHash().size(cartKey);
        return size != null ? size.intValue() : 0;
    }

    /**
     * Get cart total
     */
    public BigDecimal getCartTotal(Long userId) {
        List<CartItemDTO> items = getCartItems(userId);
        return items.stream()
            .map(item -> item.getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
            .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
}

File: frontend/src/app/services/cart.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import { tap } from 'rxjs/operators';
import { environment } from '../../environments/environment';

export interface CartItem {
  id: number;
  component: {
    id: number;
    name: string;
    price: number;
    imageUrl?: string;
  };
  quantity: number;
}

export interface CartSummary {
  items: CartItem[];
  totalItems: number;
  totalAmount: number;
}

@Injectable({
  providedIn: 'root'
})
export class CartService {
  private apiUrl = `${environment.apiUrl}/api/cart`;
  private cartItemCount$ = new BehaviorSubject<number>(0);

  constructor(private http: HttpClient) {
    // Initialize cart count on service creation
    this.refreshCartCount();
  }

  /**
   * Get observable for cart item count
   */
  getCartItemCount(): Observable<number> {
    return this.cartItemCount$.asObservable();
  }

  /**
   * Get all cart items for current user
   */
  getCartItems(): Observable<CartItem[]> {
    return this.http.get<CartItem[]>(this.apiUrl);
  }

  /**
   * Get cart summary (items + totals)
   */
  getCartSummary(): Observable<CartSummary> {
    return this.http.get<CartSummary>(`${this.apiUrl}/summary`);
  }

  /**
   * Add item to cart
   */
  addToCart(componentId: number, quantity: number = 1): Observable<CartItem> {
    return this.http.post<CartItem>(this.apiUrl, { componentId, quantity })
      .pipe(tap(() => this.refreshCartCount()));
  }

  /**
   * Update cart item quantity
   */
  updateCartItemQuantity(cartItemId: number, quantity: number): Observable<CartItem> {
    return this.http.put<CartItem>(`${this.apiUrl}/${cartItemId}`, { quantity })
      .pipe(tap(() => this.refreshCartCount()));
  }

  /**
   * Remove item from cart
   */
  removeFromCart(cartItemId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${cartItemId}`)
      .pipe(tap(() => this.refreshCartCount()));
  }

  /**
   * Clear entire cart
   */
  clearCart(): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/clear`)
      .pipe(tap(() => this.cartItemCount$.next(0)));
  }

  /**
   * Get cart total amount
   */
  getCartTotal(): Observable<{ total: number }> {
    return this.http.get<{ total: number }>(`${this.apiUrl}/total`);
  }

  /**
   * Refresh cart item count
   */
  private refreshCartCount(): void {
    this.http.get<{ count: number }>(`${this.apiUrl}/count`)
      .subscribe({
        next: (response) => this.cartItemCount$.next(response.count),
        error: () => this.cartItemCount$.next(0)
      });
  }
}
